RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc

```diff
@@ -0,0 +1,124 @@
+#include "RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithmBase.h"
+#include "RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilderPluginFactory.h"
+
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h" 
+
+GEMSegmentBuilder::GEMSegmentBuilder(const edm::ParameterSet& ps) : geom_(0) {
+  
+  // Algo name
+  algoName = ps.getParameter<std::string>("algo_name");
+  
+  edm::LogVerbatim("GEMSegmentBuilder")<< "GEMSegmentBuilder algorithm name: " << algoName;
+  
+  // SegAlgo parameter set
+  segAlgoPSet = ps.getParameter<edm::ParameterSet>("algo_pset");
+  
+  // Ask factory to build this algorithm, giving it appropriate ParameterSet  
+  algo = std::unique_ptr<GEMSegmentAlgorithmBase>(GEMSegmentBuilderPluginFactory::get()->create(algoName, segAlgoPSet));
+  
+  // Use GE21Short to have 4 rechits available in GE21?
+  useGE21Short = ps.getParameter<bool>("useGE21Short");
+
+}
+GEMSegmentBuilder::~GEMSegmentBuilder() {}
+
+void GEMSegmentBuilder::build(const GEMRecHitCollection* recHits, GEMSegmentCollection& oc) {
+  	
+  edm::LogVerbatim("GEMSegmentBuilder")<< "[GEMSegmentBuilder::build] Total number of rechits in this event: " << recHits->size();
+  
+  // Let's define the ensemble of GEM devices having the same region, chambers number (phi)
+  // different eta partitions and different layers are allowed
+
+  std::map<uint32_t, std::vector<GEMRecHit*> > ensembleRH;
+ 
+  // Loop on the GEM rechit and select the different GEM Ensemble
+  for(GEMRecHitCollection::const_iterator it2 = recHits->begin(); it2 != recHits->end(); ++it2) {
+    // GEM Ensemble is defined by assigning all the GEMDetIds of the same "superchamber" 
+    // (i.e. region same, chamber same) to the DetId of the first layer
+
+    // if use of GE21Short is prohibited and gemrechit is in station 2 quit the loop
+    if(!useGE21Short && it2->gemId().station()==2) continue;
+
+    // here a reference GEMDetId is created: named "id"
+    // - Ring 1 (no other rings available for GEM)
+    // - Layer 1 = reference layer (effective layermask)
+    // - Roll 0  = reference roll  (effective rollmask)
+    // - GE2/1 Long & Short needs to be combined => station = 2,3 ==> 3
+    // - Station == 1 (GE1/1) or == 3 (GE2/1)
+    // this reference id serves to link all GEMEtaPartitions
+    // and will also be used to determine the GEMSuperChamber 
+    // to which the GEMSegment is assigned (done inside GEMSegAlgoXX)
+    int station = 0; 
+    if(it2->gemId().station()==1) station=1;
+    else if(it2->gemId().station()==2 || it2->gemId().station()==3) station=3;
+    GEMDetId id(it2->gemId().region(),1,station,0,it2->gemId().chamber(),0);
+
+    // retrieve vector of GEMRecHits associated to the reference id
+    std::vector<GEMRecHit* > pp = ensembleRH[id.rawId()];
+    // save current GEMRecHit in vector
+    pp.push_back(it2->clone());
+    // assign updated vector of GEMRecHits to reference id
+    ensembleRH[id.rawId()]=pp;
```

Comment:
Setting aside that its not good practice to have a map of vectors, why not ensembleRH[id.rawID()].push_back(it2->clone()) instead of these last three lines